File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TcImportsHack.fs
﻿namespace JetBrains.ReSharper.Plugins.FSharp.Shim.TypeProviders

open System.Reflection
open JetBrains.Rider.FSharp.TypeProviders.Protocol

module TcImportsHack =
    let bindAll =
        BindingFlags.DeclaredOnly ||| BindingFlags.Public ||| BindingFlags.NonPublic |||
        BindingFlags.Static ||| BindingFlags.Instance

    type System.Object with
        member x.GetProperty(nm) =
            let ty = x.GetType()
            let prop = ty.GetProperty(nm, bindAll)
            let v = prop.GetValue(x, null)
            v

        member x.GetField(nm) =
            let ty = x.GetType()
            let fld = ty.GetField(nm, bindAll)
            let v = fld.GetValue(x)
            v

    type FakeDllInfo(fileName: string) =
        member this.FileName = fileName

    type FakeTcImportsBaseValue(baseFakeTcImports: Server.RdFakeTcImports) =
        member this.Value = FakeTcImports(baseFakeTcImports.DllInfos, baseFakeTcImports.Base)

    and FakeTcImports(dllInfos: Server.RdFakeDllInfo[], baseFakeTcImports: Server.RdFakeTcImports) as this =
        [<DefaultValue>] val mutable dllInfos: Server.RdFakeDllInfo[]
        do this.dllInfos <- dllInfos

        member this.Base = FakeTcImportsBaseValue(baseFakeTcImports)
        member this.SystemRuntimeContainsType(_: string) = true // todo: smart implementation

    type FakeSystemRuntimeContainsTypeRef(fakeTcImports: Server.RdFakeTcImports) =
        member this.Value =
            let tcImports = FakeTcImports(fakeTcImports.DllInfos, fakeTcImports.Base)
            // Don't simplify the lambda: `tcImports` is accessed via reflection in `getFakeTcImports`.
            fun x -> tcImports.SystemRuntimeContainsType x

    // The type provider must not contain strong references to remote TcImport objects.
    // The legacy Type Provider SDK gets dllInfos data from the 'systemRuntimeContainsType' closure.
    // This hack allows you to pull this data for transfer between processes.
    let getFakeTcImports (runtimeContainsType: 'a -> 'b) =
        let getDllInfos tcImports =
            [| for dllInfo in (tcImports.GetField("dllInfos")  :?> _) ->
                   Client.RdFakeDllInfo(dllInfo.GetProperty("FileName") :?> _) |]

        let tcImports =
            runtimeContainsType.GetField("systemRuntimeContainsTypeRef").GetProperty("Value").GetField("tcImports")

        let tcImportsDllInfos = getDllInfos tcImports

        let baseTcImports = tcImports.GetProperty("Base").GetProperty("Value")
        let baseTcImportsDllInfos = getDllInfos baseTcImports

        let fakeBaseTcImports = Client.RdFakeTcImports(null, baseTcImportsDllInfos)
        Client.RdFakeTcImports(fakeBaseTcImports, tcImportsDllInfos)

    let injectFakeTcImports (fakeTcImports: Server.RdFakeTcImports) =
        let systemRuntimeContainsTypeRef = FakeSystemRuntimeContainsTypeRef(fakeTcImports)
        // Don't simplify the lambda: `systemRuntimeContainsTypeRef` is accessed via reflection in `getFakeTcImports`.
        fun name -> systemRuntimeContainsTypeRef.Value name

File: ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/TcImportsHack.cs
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using JetBrains.Rider.FSharp.TypeProviders.Protocol.Server;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Core;
using ClientRdFakeTcImports = JetBrains.Rider.FSharp.TypeProviders.Protocol.Client.RdFakeTcImports;
using ClientRdFakeDllInfo = JetBrains.Rider.FSharp.TypeProviders.Protocol.Client.RdFakeDllInfo;

namespace JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol
{
  internal static class ReflectionUtils
  {
    private const BindingFlags BIND_ALL = BindingFlags.DeclaredOnly | BindingFlags.Public | BindingFlags.NonPublic |
                                          BindingFlags.Static | BindingFlags.Instance;

    public static object GetProperty(this object obj, string nm)
    {
      var ty = obj.GetType();
      var prop = ty.GetProperty(nm, BIND_ALL);
      var v = prop.GetValue(obj, null);
      return v;
    }

    public static object GetField(this object obj, string nm)
    {
      var ty = obj.GetType();
      var fld = ty.GetField(nm, BIND_ALL);
      var v = fld.GetValue(obj);
      return v;
    }
  }

  public static class TcImportsHack
  {
    record FakeTcImportsBaseValue(RdFakeTcImports baseFakeTcImports)
    {
      public FakeTcImports Value => new(baseFakeTcImports.DllInfos, baseFakeTcImports.Base);
    }

    record FakeTcImports(RdFakeDllInfo[] DllInfos, RdFakeTcImports baseFakeTcImports)
    {
      public RdFakeDllInfo[] dllInfos = DllInfos;
      public FakeTcImportsBaseValue Base => new(baseFakeTcImports);
      public bool SystemRuntimeContainsType(string _) => true; // todo: smart implementation
    }

    record FakeSystemRuntimeContainsTypeRef(RdFakeTcImports fakeTcImports)
    {
      public FSharpFunc<string, bool> Value =>
        new FakeTcImportsClosure(new FakeTcImports(fakeTcImports.DllInfos, fakeTcImports.Base));
    }

    private sealed class FakeTcImportsClosure : FSharpFunc<string, bool>
    {
      private readonly FakeTcImports tcImports;

      public FakeTcImportsClosure(FakeTcImports tcImports) => this.tcImports = tcImports;

      public override bool Invoke(string x) => tcImports.SystemRuntimeContainsType(x);
    }

    private sealed class SystemRuntimeContainsTypeRefClosure : FSharpFunc<string, bool>
    {
      private readonly FakeSystemRuntimeContainsTypeRef systemRuntimeContainsTypeRef;

      public SystemRuntimeContainsTypeRefClosure(FakeSystemRuntimeContainsTypeRef systemRuntimeContainsTypeRef) =>
        this.systemRuntimeContainsTypeRef = systemRuntimeContainsTypeRef;

      public override bool Invoke(string x) => systemRuntimeContainsTypeRef.Value.Invoke(x);
    }


    // The type provider must not contain strong references to remote TcImport objects.
    // The legacy Type Provider SDK gets dllInfos data from the 'systemRuntimeContainsType' closure.
    // This hack allows you to pull this data for transfer between processes.
    public static ClientRdFakeTcImports GetFakeTcImports(FSharpFunc<string, bool> runtimeContainsType)
    {
      ClientRdFakeDllInfo[] getDllInfos(object imports) => (imports.GetField("dllInfos") as IEnumerable<object>)
        .Select(dllInfo => new ClientRdFakeDllInfo(dllInfo.GetProperty("FileName") as string))
        .ToArray();

      var tcImports =
        runtimeContainsType.GetField("systemRuntimeContainsTypeRef").GetProperty("Value").GetField("tcImports");

      var tcImportsDllInfos = getDllInfos(tcImports);
      var baseTcImports = tcImports.GetProperty("Base").GetProperty("Value");
      var baseTcImportsDllInfos = getDllInfos(baseTcImports);
      var fakeBaseTcImports = new ClientRdFakeTcImports(null, baseTcImportsDllInfos);

      return new ClientRdFakeTcImports(fakeBaseTcImports, tcImportsDllInfos);
    }

    public static FSharpFunc<string, bool> InjectFakeTcImports(RdFakeTcImports fakeTcImports) =>
      new SystemRuntimeContainsTypeRefClosure(new FakeSystemRuntimeContainsTypeRef(fakeTcImports));
  }
}
